<a href="https://colab.research.google.com/github/jccamposro/AlgorithmsUN2023I/blob/main/Lab4/jccamposroGroup6Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>
using namespace std;

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    cin >> a;
    cin >> b;
    cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB
!./APlusB



8989 6565
15554

### Test input output files

In [ ]:
%%writefile input1.txt
3 23

Writing input1.txt


In [ ]:
%%writefile output1.txt
26

Writing output1.txt


In [ ]:
%%writefile input2.txt
-8 3

Writing input2.txt


In [ ]:
%%writefile output2.txt
7

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

26

In [ ]:
!./APlusB < input2.txt

-5

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt 

In [ ]:
!grep -xvFf output1.txt output2.txt 

7


In [ ]:
!grep -xvFf output2.txt output1.txt 

26


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

-5


## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

5 8
13


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt


-5


# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int MaxPairwiseProduct(const vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = max(max_product, numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    cin >> n;
    vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        cin >> numbers[i];
    }

    cout << "\n" << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [ ]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

5
45 48 69 2222 5

153318


### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt


90


### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [ ]:
!./max_pairwise_product < input2.txt


1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    cin >> n;
    vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        cin >> numbers[i];
    }

    cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [ ]:
%%writefile max_pairwise_product_long_test3.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    srand(time(NULL));
    
    for(int i = 0; i < 20; i++){
      int n = rand() % 19 + 2;
      vector<int> nums(n);

      for(int j = 0; j < n; j++){
        nums[j] = rand() % LLONG_MAX + 1;
      }

      cout << "Test " << i << ": ";
      for(int j = 0; j < n; j++){
        cout << nums[j] << " ";
      }

      cout << "\n" << "Maximum pairwise product is: " << MaxPairwiseProduct(nums) << endl;

    }
    return 0;
}

Writing max_pairwise_product_long_test3.cpp


In [ ]:
!g++ max_pairwise_product_long_test3.cpp -o max_pairwise_product_long_test3
!./max_pairwise_product_long_test3

Test 0: 496984427 566346184 48147508 2101711069 1755084220 2071428632 2011866101 486722092 1189685025 1994340068 234982021 30935150 977392401 1856792251 1589302704 700859853 1395384351 1522986981 
Maximum pairwise product is: 4353544484517927608
Test 1: 1059962029 1338590300 1329634627 
Maximum pairwise product is: 1779836014246318100
Test 2: 505068201 1583572457 1805153912 940734980 675051667 2147466876 
Maximum pairwise product is: 3876508232101818912
Test 3: 1689746323 496967654 1106876410 1737893830 451195074 714476981 
Maximum pairwise product is: 2936599709006887090
Test 4: 315577526 1201199072 704040190 162433945 1436181092 734975339 1139826345 1145489694 176794394 1840686197 393390396 1699781374 1924743358 1453352425 
Maximum pairwise product is: 3542848531838029526
Test 5: 1106894336 602876820 1395956225 542983144 260547084 189207556 1218034810 260530311 729737782 
Maximum pairwise product is: 1700323275286192250
Test 6: 757497964 1836614191 
Maximum pairwise product is: 13912

### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [ ]:
%%writefile Test2Write.cpp
#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Writing Test2Write.cpp


In [ ]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [ ]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {                                                               
    int n = 10;                                                                                                                            //1
    vector<int> numbers(n);                                            
    for (int i = 0; i < n; ++i) {                                           
        cout << numbers[i];                                             
    }

    cout << "\n";                       
    return 0;                                                              
}  

Writing test_init_vectors.cpp


In [ ]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers) {                   //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    cin >> n;                                             //1
    vector<int> numbers(n);                               //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        cin >> numbers[i];                                //n
    }

    cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [ ]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                //N inst exec by line executed 
    long long execints = 0LL;

    int n;                                                  //1
    execints++;                                                              
    
    cin >> n;                                               //1
    execints++;

    vector<int> numbers(n);                                 //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        cin >> numbers[i];
        execints++;                                          //n
    }
    
    //cout << MaxPairwiseProduct(numbers) << "\n";          //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Writing max_pairwise_product_long_execints.cpp


In [ ]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
5656
555
^C


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [ ]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [ ]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile efficient_maximum_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int EfficientMaximumPairwiseProduct(int n, vector<long long> nums){
    int maxn = 0;                                                               //1
    int smaxn = 0;                                                              //1
    for(int i = 0; i < n; i++){                                                 //n
      if(nums[i] > maxn){                                                       //n
        smaxn = maxn;                                                           //n
        maxn = nums[i];                                                         //n
      }else if(nums[i] > smaxn){                                                //n
        smaxn = nums[i];                                                        //n
      }
    }
    return smaxn*maxn;                                                          //1
}                                                                               //O(6n+3)

int main () {
    int n;
    cin >> n;
    vector<long long> nums(n);
    for(int i = 0; i < n; i++){                                                 //n
        cin >> nums[i];                                                         //n
    }

    cout << "Maximum pairwise product is: " << EfficientMaximumPairwiseProduct(n, nums) << endl;
}

Writing efficient_maximum_pairwise_product.cpp


Test with inputs 1 and 2



In [ ]:
!g++ efficient_maximum_pairwise_product.cpp -o efficient_maximum_pairwise_product
!time ./efficient_maximum_pairwise_product < input1.txt

!g++ efficient_maximum_pairwise_product.cpp -o efficient_maximum_pairwise_product
!time ./efficient_maximum_pairwise_product < input2.txt
     

Maximum pairwise product is: 90

real	0m0.002s
user	0m0.002s
sys	0m0.000s
Maximum pairwise product is: 1345294336

real	0m0.002s
user	0m0.002s
sys	0m0.000s


Test and time it with input 3

In [ ]:
!g++ efficient_maximum_pairwise_product.cpp -o efficient_maximum_pairwise_product
!time ./efficient_maximum_pairwise_product < input3.txt

Maximum pairwise product is: 1345094336

real	0m0.066s
user	0m0.060s
sys	0m0.005s


20 stress tests

In [ ]:
%%writefile efficient_maximum_pairwise_product_stress.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

int EfficientMaximumPairwiseProduct(int n, vector<long long> nums){
    int maxn = 0;                                                               //1
    int smaxn = 0;                                                              //1
    for(int i = 0; i < n; i++){                                                 //n
      if(nums[i] > maxn){                                                       //n
        smaxn = maxn;                                                           //n
        maxn = nums[i];                                                         //n
      }else if(nums[i] > smaxn){                                                //n
        smaxn = nums[i];                                                        //n
      }
    }
    return smaxn*maxn;                                                          //1
}                                                                               //O(6n+3)

int main() {
    srand(time(NULL));

    for(int i = 0; i < 20; i++){
      int n = rand() % 19 + 2;
      vector<long long> nums(n);

      for(int j = 0; j < n; j++){
        nums[j] = rand() % LLONG_MAX + 1;
      }

      cout << "Test " << i << ": ";
      for(int j = 0; j < n; j++){
        cout << nums[j] << " ";
      }

      cout << "\n" << "Maximum pairwise product is: " << EfficientMaximumPairwiseProduct(n, nums) << endl;

    }
    return 0;
}

Writing efficient_maximum_pairwise_product_stress.cpp


In [ ]:
!g++ efficient_maximum_pairwise_product_stress.cpp -o efficient_maximum_pairwise_product_stress
!time ./efficient_maximum_pairwise_product_stress

Test 0: 1839745509 967081131 
Maximum pairwise product is: 384321527
Test 1: 251737745 1584315994 1363568579 803078858 757065910 640392038 1838094661 1173539663 1083312609 821520437 
Maximum pairwise product is: -357164990
Test 2: 296305487 1313446186 219005156 243800308 1265287424 934032469 1594513864 1814384384 176000362 160935410 1992243686 
Maximum pairwise product is: 1763441152
Test 3: 1715778228 972420127 1557308011 1749216789 1518675429 1249569871 568814272 1075274182 1501307616 5646617 291359113 156902825 762712527 931751150 1994997485 1936252189 
Maximum pairwise product is: -620316199
Test 4: 669034274 908369428 163885597 1982480459 1127374583 407685904 1100284234 
Maximum pairwise product is: -1241148835
Test 5: 2002199767 767184969 89923765 15651528 611945007 361845604 1731429755 1584365133 1919153614 1333162896 955556913 1021239837 1901977167 2030831095 375063804 1907623783 174706559 531966628 522852661 1106457708 
Maximum pairwise product is: 836812401
Test 6: 311621202 

Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile efficient_maximum_pairwise_product_exec.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int EfficientMaximumPairwiseProduct(int n, vector<long long> nums){
    int maxn = 0;                                                               //1
    int smaxn = 0;                                                              //1
    for(int i = 0; i < n; i++){                                                 //n
      if(nums[i] > maxn){                                                       //n
        smaxn = maxn;                                                           //n
        maxn = nums[i];                                                         //n
      }else if(nums[i] > smaxn){                                                //n
        smaxn = nums[i];                                                        //n
      }
    }
    return smaxn*maxn;                                                          //1
}                                                                               //O(6n+3)

int main () {
    int n;
    cin >> n;
    vector<long long> nums(n);
    for(int i = 0; i < n; i++){                                                 //n
        cin >> nums[i];                                                         //n
    }

    cout << "\n" << "Instruction number: " << 8*n+3 << endl;
    cout << "Maximum pairwise product is: " << EfficientMaximumPairwiseProduct(n, nums) << endl;
}

Writing efficient_maximum_pairwise_product_exec.cpp


In [ ]:
!g++ efficient_maximum_pairwise_product_exec.cpp -o efficient_maximum_pairwise_product_exec
!time ./efficient_maximum_pairwise_product_exec < input1.txt

!g++ efficient_maximum_pairwise_product_exec.cpp -o efficient_maximum_pairwise_product_exec
!time ./efficient_maximum_pairwise_product_exec < input2.txt

!g++ efficient_maximum_pairwise_product_exec.cpp -o efficient_maximum_pairwise_product_exec
!time ./efficient_maximum_pairwise_product_exec < input3.txt


Instruction number: 83
Maximum pairwise product is: 90

real	0m0.002s
user	0m0.001s
sys	0m0.001s

Instruction number: 51
Maximum pairwise product is: 1345294336

real	0m0.003s
user	0m0.001s
sys	0m0.002s

Instruction number: 1600003
Maximum pairwise product is: 1345094336

real	0m0.060s
user	0m0.056s
sys	0m0.003s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
#Estimated time by instruction and number of instructions by second
print("Estimated time by instruction", "\t \t", "Number of instructions by second")
print("Input 1: ", (83/0.003), "\t \t", 1/(83/0.003))
print("Input 2: ", (51/0.001), "\t \t \t", 1/(51/0.001))
print("Input 3: ", (43/0.002), "\t \t \t", 1/(43/0.002))

Estimated time by instruction 	 	 Number of instructions by second
Input 1:  27666.666666666668 	 	 3.614457831325301e-05
Input 2:  51000.0 	 	 	 1.9607843137254903e-05
Input 3:  21500.0 	 	 	 4.651162790697674e-05


### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [ ]:
%%writefile super_efficient_maximum_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main () {
    int n;                                                                      //1
    cin >> n;                                                                   //1
    int maxn = 0;                                                               //1
    int smaxn = 0;                                                              //1
    for(int i = 0; i < n; i++){                                                 //n
      int num;                                                                  //n
      cin >> num;                                                               //n
      if(num > maxn){                                                           //n
        smaxn = maxn;                                                           //n
        maxn = num;                                                             //n
      }else if(num > smaxn){                                                    //n
        smaxn = num;                                                            //n
      }
    }
    cout << "Maximum pairwise product is: " << smaxn*maxn << endl;              //1
}                                                                               //O(8*n+4)

Writing super_efficient_maximum_pairwise_product.cpp


Test with inputs 1 and 2


 

In [ ]:
!g++ super_efficient_maximum_pairwise_product.cpp -o super_efficient_maximum_pairwise_product
!time ./super_efficient_maximum_pairwise_product < input1.txt

!g++ super_efficient_maximum_pairwise_product.cpp -o super_efficient_maximum_pairwise_product
!time ./super_efficient_maximum_pairwise_product < input2.txt

Maximum pairwise product is: 90

real	0m0.003s
user	0m0.003s
sys	0m0.000s
Maximum pairwise product is: 1345294336

real	0m0.005s
user	0m0.002s
sys	0m0.000s


Test and time it with input 3

In [ ]:
!g++ super_efficient_maximum_pairwise_product.cpp -o super_efficient_maximum_pairwise_product
!time ./super_efficient_maximum_pairwise_product < input3.txt

Maximum pairwise product is: 1345094336

real	0m0.054s
user	0m0.052s
sys	0m0.001s


20 stress tests 

In [ ]:
%%writefile super_efficient_maximum_pairwise_product_stress.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

int main () {
    srand(time(NULL));                                                          //1
    for(int i = 0; i < 20; i++){                                                //t = 20
        int maxn = 0;                                                               //1
        int smaxn = 0;                                                              //1
        int n = rand() % 19 + 2;                                                    //t
        for(int j = 0; j < n; j++){                                                   //t*n
          int num;                                                                    //t*n
          num = rand() % LLONG_MAX + 1;                                               //t*n
          if(num > maxn){                                                             //t*n
            smaxn = maxn;                                                             //t*n
            maxn = num;                                                               //t*n
          }else if(num > smaxn){                                                      //t*n
            smaxn = num;                                                              //t*n
          }
        }
        cout << "Maximum pairwise product is: " << smaxn*maxn << endl;              //1
    }
    return 0;
}                                                                               //O(9*n+4)

Writing super_efficient_maximum_pairwise_product_stress.cpp


In [ ]:
!g++ super_efficient_maximum_pairwise_product_stress.cpp -o super_efficient_maximum_pairwise_product_stress
!time ./super_efficient_maximum_pairwise_product_stress

Maximum pairwise product is: -424906816
Maximum pairwise product is: -1122688640
Maximum pairwise product is: 616597554
Maximum pairwise product is: 269756580
Maximum pairwise product is: 1340954880
Maximum pairwise product is: 1201715887
Maximum pairwise product is: -1678879936
Maximum pairwise product is: 317142814
Maximum pairwise product is: 2042597876
Maximum pairwise product is: 115673762
Maximum pairwise product is: -1317852885
Maximum pairwise product is: -306621586
Maximum pairwise product is: -323927172
Maximum pairwise product is: 1381523944
Maximum pairwise product is: -1954369870
Maximum pairwise product is: -214944010
Maximum pairwise product is: -1283857888
Maximum pairwise product is: -602992444
Maximum pairwise product is: -1612530720
Maximum pairwise product is: -1259426898

real	0m0.003s
user	0m0.001s
sys	0m0.001s


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile super_efficient_maximum_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main () {
    int n;                                                                      //1
    cin >> n;                                                                   //1
    int maxn = 0;                                                               //1
    int smaxn = 0;                                                              //1
    for(int i = 0; i < n; i++){                                                 //n
      int num;                                                                  //n
      cin >> num;                                                               //n
      if(num > maxn){                                                           //n
        smaxn = maxn;                                                           //n
        maxn = num;                                                             //n
      }else if(num > smaxn){                                                    //n
        smaxn = num;                                                            //n
      }
    }
    cout << "\n" << "Instruction number: " << 8*n+3 << endl;                    //1
    cout << "Maximum pairwise product is: " << smaxn*maxn << endl;              //1
}                                                                               //O(8*n+6)

Overwriting super_efficient_maximum_pairwise_product.cpp


In [ ]:
!g++ super_efficient_maximum_pairwise_product.cpp -o super_efficient_maximum_pairwise_product
!time ./super_efficient_maximum_pairwise_product < input1.txt

!g++ super_efficient_maximum_pairwise_product.cpp -o super_efficient_maximum_pairwise_product
!time ./super_efficient_maximum_pairwise_product < input2.txt

!g++ super_efficient_maximum_pairwise_product.cpp -o super_efficient_maximum_pairwise_product
!time ./super_efficient_maximum_pairwise_product < input3.txt


Instruction number: 83
Maximum pairwise product is: 90

real	0m0.002s
user	0m0.000s
sys	0m0.002s

Instruction number: 51
Maximum pairwise product is: 1345294336

real	0m0.003s
user	0m0.002s
sys	0m0.000s

Instruction number: 1600003
Maximum pairwise product is: 1345094336

real	0m0.097s
user	0m0.092s
sys	0m0.001s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
#Estimated time by instruction and number of instructions by second
print("Estimated time by instruction", "\t \t", "Number of instructions by second")
print("Input 1: ", (83/0.002), "\t \t \t", 1/(83/0.002))
print("Input 2: ", (51/0.003), "\t \t \t", 1/(51/0.003))
print("Input 3: ", (43/0.004), "\t \t \t", 1/(43/0.004))

Estimated time by instruction 	 	 Number of instructions by second
Input 1:  41500.0 	 	 	 2.4096385542168674e-05
Input 2:  17000.0 	 	 	 5.882352941176471e-05
Input 3:  10750.0 	 	 	 9.302325581395348e-05


# Codeforces problem https://codeforces.com/contest/1619/problem/A 

Solve the problem in Colab, submit the solution to Codeforces and add a screenshoot of the submission.

My profile: 




Submitted code:


Submission proof:
